supreme/src/androidMain/kotlin/at/asitplus/signum/supreme/os/AndroidKeyStoreProvider.kt

```diff
@@ -0,0 +1,393 @@
+            } ?: throw UnsupportedCryptoException("Unsupported $what $only")
+        }
+    }
+
+private fun attestationFor(chain: CertificateChain) =
+    if (chain.size > 1) AndroidKeystoreAttestation(chain) else null
+
+/**
+ * A provider that manages keys in the [Android Key Store](https://developer.android.com/privacy-and-security/keystore).
+ */
+object AndroidKeyStoreProvider:
+    PlatformSigningProviderI<AndroidKeystoreSigner, AndroidSignerConfiguration, AndroidSigningKeyConfiguration>
+{
+
+    private val ks: KeyStore get() =
+        KeyStore.getInstance("AndroidKeyStore").apply { load(null, null) }
+
+    @SuppressLint("WrongConstant")
+    override suspend fun createSigningKey(
+        alias: String,
+        configure: DSLConfigureFn<AndroidSigningKeyConfiguration>
+    ) = catching {
+        if (ks.containsAlias(alias)) {
+            throw NoSuchElementException("Key with alias $alias already exists")
+        }
+        val config = DSL.resolve(::AndroidSigningKeyConfiguration, configure)
+        val spec = KeyGenParameterSpec.Builder(
+            alias,
+            KeyProperties.PURPOSE_SIGN
+        ).apply {
+            when(val algSpec = config._algSpecific.v) {
+                is SigningKeyConfiguration.RSAConfiguration -> {
+                    setAlgorithmParameterSpec(
+                        RSAKeyGenParameterSpec(algSpec.bits, algSpec.publicExponent.toJavaBigInteger()))
+                    setDigests(*algSpec.digests.map(Digest::jcaName).toTypedArray())
+                    setSignaturePaddings(*algSpec.paddings.map {
+                        when (it) {
+                            RSAPadding.PKCS1 -> KeyProperties.SIGNATURE_PADDING_RSA_PKCS1
+                            RSAPadding.PSS -> KeyProperties.SIGNATURE_PADDING_RSA_PSS
+                        }
+                    }.toTypedArray())
+                }
+                is SigningKeyConfiguration.ECConfiguration -> {
+                    setAlgorithmParameterSpec(ECGenParameterSpec(algSpec.curve.jcaName))
+                    setDigests(*algSpec.digests.map { it?.jcaName ?: KeyProperties.DIGEST_NONE }.toTypedArray())
+                }
+            }
+            setCertificateNotBefore(Date.from(Instant.now()))
+            setCertificateSubject(X500Principal("CN=$alias")) // TODO
+            config.hardware.v?.let { hw ->
+                setIsStrongBoxBacked(when (hw.strongBox) {
+                    REQUIRED -> true
+                    PREFERRED -> false // TODO
+                    DISCOURAGED -> false
+                })
+                hw.attestation.v?.let {
+                    setAttestationChallenge(it.challenge)
+                }
+                hw.protection.v?.let {
+                    setUserAuthenticationRequired(true)
+                    setUserAuthenticationParameters(it.timeout.inWholeSeconds.toInt(),
+                        (if (it.factors.v.biometry) KeyProperties.AUTH_BIOMETRIC_STRONG else 0) or
```

Comment:
note to self: may need switch depending on android version. this depends on target and minsdk etc.